model/map/src/main/java/org/keycloak/models/map/role/MapRoleProvider.java

```diff
@@ -0,0 +1,344 @@
+
+public class MapRoleProvider implements RoleProvider {
+
+    private static final Logger LOG = Logger.getLogger(MapRoleProvider.class);
+    private static final Predicate<MapRoleEntity> ALWAYS_FALSE = role -> { return false; };
+    private final KeycloakSession session;
+    final MapKeycloakTransaction<UUID, MapRoleEntity> tx;
+    private final MapStorage<UUID, MapRoleEntity> roleStore;
+
+    private static final Comparator<MapRoleEntity> COMPARE_BY_NAME = new Comparator<MapRoleEntity>() {
+        @Override
+        public int compare(MapRoleEntity o1, MapRoleEntity o2) {
+            String r1 = o1 == null ? null : o1.getName();
+            String r2 = o2 == null ? null : o2.getName();
+            return r1 == r2 ? 0
+              : r1 == null ? -1
+              : r2 == null ? 1
+              : r1.compareTo(r2);
+
+        }
+    };
+
+    public MapRoleProvider(KeycloakSession session, MapStorage<UUID, MapRoleEntity> roleStore) {
+        this.session = session;
+        this.roleStore = roleStore;
+        this.tx = new MapKeycloakTransaction<>(roleStore);
+        session.getTransactionManager().enlist(tx);
+    }
+
+    private Function<MapRoleEntity, RoleModel> entityToAdapterFunc(RealmModel realm) {
+        // Clone entity before returning back, to avoid giving away a reference to the live object to the caller
+
+        return origEntity -> new MapRoleAdapter(session, realm, registerEntityForChanges(origEntity));
+    }
+
+    private MapRoleEntity registerEntityForChanges(MapRoleEntity origEntity) {
+        final MapRoleEntity res = Serialization.from(origEntity);
+        tx.putIfChanged(origEntity.getId(), res, MapRoleEntity::isUpdated);
+        return res;
+    }
+
+    private Predicate<MapRoleEntity> entityRealmFilter(RealmModel realm) {
+        if (realm == null || realm.getId() == null) {
+            return MapRoleProvider.ALWAYS_FALSE;
+        }
+        String realmId = realm.getId();
+        return entity -> Objects.equals(realmId, entity.getRealmId());
+    }
+
+    private Predicate<MapRoleEntity> entityClientFilter(ClientModel client) {
+        if (client == null || client.getId() == null) {
+            return MapRoleProvider.ALWAYS_FALSE;
+        }
+        String clientId = client.getId();
+        return entity -> entity.isClientRole() && 
+                Objects.equals(clientId, entity.getClientId());
+    }
+
+    private Stream<MapRoleEntity> getNotRemovedUpdatedRolesStream() {
+        Stream<MapRoleEntity> updatedAndNotRemovedRolesStream = roleStore.entrySet().stream()
+          .map(tx::getUpdated)    // If the role has been removed, tx.get will return null, otherwise it will return me.getValue()
+          .filter(Objects::nonNull);
+        return Stream.concat(tx.createdValuesStream(roleStore.keySet()), updatedAndNotRemovedRolesStream);
+    }
+
+    @Override
+    public RoleModel addRealmRole(RealmModel realm, String id, String name) {
```

Comment:
Here we should check that there is no realm role with the name for the realm. This is enforced [here](https://github.com/keycloak/keycloak/blob/master/model/jpa/src/main/java/org/keycloak/models/jpa/entities/RoleEntity.java#L55) for the jpa layer.